FEATURE: admin can disable flags

lis2 opened the pull request:
UI for admins to disable system flags.

<img width="1441" alt="Screenshot 2024-05-24 at 2 22 08 PM" src="https://github.com/discourse/discourse/assets/72780/20edb120-ee6a-465d-b2f6-a67868ddb9ba">

martin-brennan commented on app/assets/javascripts/admin/addon/components/admin-flag.gjs:
```diff
@@ -0,0 +1,41 @@
+
+    return ajax(`/admin/flags/${flag.id}/toggle`, {
+      type: "PUT",
+      contentType: "application/json",
```
I don't _think_ you need to specify this, IIRC it's default

martin-brennan commented on app/assets/javascripts/admin/addon/components/admin-flag.gjs:
```diff
@@ -0,0 +1,41 @@
+  }
+
+  <template>
+    <tr class="flag">
```
We should aim for more specificity with CSS classes -- something like `admin-flag-row` in this case would be good. Then you have e.g. `admin-flag-row__name`. This avoids conflicts with other CSS, `flag` is quite a generic class name.

martin-brennan commented on app/assets/javascripts/admin/addon/components/admin-flags.gjs:
```diff
@@ -0,0 +1,27 @@
+  @tracked flags = this.site.flagTypes;
+
+  <template>
+    <div class="container flags">
```
Same comment about CSS names -- `admin-flags` would be better here for example.

martin-brennan commented on app/assets/javascripts/admin/addon/components/admin-flags.gjs:
```diff
@@ -0,0 +1,27 @@
+
+export default class AdminFlags extends Component {
+  @service site;
+  @tracked flags = this.site.flagTypes;
```
I don't think this needs to be tracked since the reactivity is limited to the sub-component, try doing `flags = this.site.flagTypes` here

martin-brennan commented on app/assets/javascripts/admin/addon/routes/admin-route-map.js:
```diff
@@ -209,6 +209,14 @@ export default function () {
     );
 
+    this.route(
+      "adminFlags",
```
I think this fine for now but pending questions around config area we may move it to /admin/config/:area route, shouldn't be too hard to recalibrate at this point :)

martin-brennan commented on app/assets/javascripts/admin/addon/templates/flags-index.hbs:
```diff
@@ -0,0 +1 @@
+<AdminFlags />
```
Love to see it. I hate going back to old routes and seeing 100000 things in the route template :grimacing:

martin-brennan commented on app/models/post_action_type.rb:
```diff
@@ -67,7 +68,16 @@ def flag_types_without_custom
       return flag_settings.flag_types if overridden_by_plugin_or_skipped_db?
-      flag_enum(all_flags)
+
+      # Once replace_flag API is fully deprecated, then we can drop respond_to. It is needed right now for migration to be evaluated.
```
Please put this + the other comment as a `TODO (lis2/krisk)` so we know who is responsible :)

martin-brennan commented on app/models/reviewable_score.rb:
```diff
@@ -11,7 +11,7 @@ class ReviewableScore < ActiveRecord::Base
   # compatibility, but we can add extra reasons for scores.
   def self.types
-    @types ||= PostActionType.flag_types.merge(needs_approval: 9)
+    @types ||= PostActionType.flag_types.merge(needs_approval: 9).merge(PostActionType.score_types)
```
Will we be able to get rid of this weird `merge(needs_approval: 9)` at some point?

lis2 replied:
oh yes! since we have score type I can add it to fixtures and have that in the database!

martin-brennan commented on app/serializers/post_action_type_serializer.rb:
```diff
@@ -29,6 +38,10 @@ def name_key
   end
 
+  def enabled
+    !PostActionType.disabled_flag_types[object.id]
```
Might be nice to add an `enabled_flag_types` shortcut method to avoid the double negative here

martin-brennan commented on app/serializers/toggle_flag.rb:
```diff
@@ -0,0 +1,39 @@
+  end
+
+  class Contract
+    attribute :flag_id, :integer
```
You could add a validates presence for the `flag_id` too

https://github.com/discourse/discourse/blob/d1191b7f5fbe9e8ea1b645d5a61c79fa93693c0c/plugins/chat/app/services/chat/create_thread.rb#L38-L38

martin-brennan commented on spec/lib/guardian/flag_guardian_spec.rb:
```diff
@@ -32,4 +32,11 @@
+  describe "#can_toggle_flag?" do
+    it "returns true for admin and false for regular user" do
+      expect(Guardian.new(admin).can_toggle_flag?).to eq(true)
+      expect(Guardian.new(user).can_toggle_flag?).to eq(false)
```
Would just add one more check for a `moderator` too so it's intentionally showing that we aren't allowing mods to do this

martin-brennan commented on spec/services/toggle_flag_spec.rb:
```diff
@@ -0,0 +1,45 @@
+
+  let(:flag) { Flag.system.last }
+
+  let(:guardian) { Guardian.new(current_user) }
```
No need for this, you can do `current_user.guardian`

martin-brennan commented on spec/system/page_objects/pages/admin_flags.rb:
```diff
@@ -0,0 +1,11 @@
+  module Pages
+    class AdminFlags < PageObjects::Pages::Base
+      def toggle(key)
+        PageObjects::Components::DToggleSwitch.new(".#{key}").toggle
```
Maybe scope this a bit further e.g. `.admin-flags-row.#{key}` or however the class is structured

martin-brennan commented on app/controllers/admin/flags_controller.rb:
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+class Admin::FlagsController < Admin::StaffController
```
I think this could be `Admin::AdminController` for now, I don't think mods will be doing this. Later on if they do we could change it back.

martin-brennan commented on app/controllers/admin/flags_controller.rb:
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+class Admin::FlagsController < Admin::StaffController
+  include WithServiceHelper
```
Let's put this in `AdminController` and `StaffController` base classes so we can use it globally :fire:

martin-brennan commented on app/assets/javascripts/discourse/app/lib/sidebar/admin-nav-map.js:
```diff
@@ -110,6 +110,12 @@ export const ADMIN_NAV_MAP = [
         icon: "gavel",
       },
+      {
+        name: "admin_moderation_flags",
```
I do wonder if for now we should hide this behind a `admin_flagging_ui_enabled_groups` feature flag? I only worry a little about admins going to this page since it's new then being a bit confused about why they would want to enable/disable flags.

Up to you, if you think it won't be a problem then leave it :+1:

lis2 replied:
I feel quite confident here. The feature may not be the most useful but still can be used. Even UI is reflecting that if all flags are disabled it shows information:

<img width="626" alt="Screenshot 2024-05-29 at 11 31 34 AM" src="https://github.com/discourse/discourse/assets/72780/2af55d4f-ec0d-476b-b1a5-62c11b725c23">

martin-brennan commented on app/assets/javascripts/admin/addon/components/admin-flag.gjs:
```diff
@@ -0,0 +1,41 @@
+    <tr class="flag">
+      <td>
+        <p class="flag__name">{{@flag.name}}</p>
+        <p class="flag__description">{{@flag.description}}</p>
```
Do we need to do `htmlSafe` on the description? It seems like it in the screenshot.

martin-brennan reviewed (commented):
Looks great, I'm super happy that you've used the Service class and gotten to try it out finally. I wouldn't have thought to use it for a Toggle, but it actually works nicely and it's so clear what that class is doing :ok_hand: 

I do have one question about disabled flags -- if you do a PUT request to flag a topic/post outside the UI, do we raise an error if that flag is disabled?

lis2 commented:
> - if you do a PUT request to flag a topic/post outside the UI, do we raise an error if that flag is disabled?

Good catch. I just added check for post_can_act guardian to ensure that flag is enabled and added spec

martin-brennan reviewed (approved):
Just need to fix up the last few comments too, thanks Kris!

forkfork987 commented:
@lis2 I got this error when sync my fork repo
`fatal: cannot create directory at 'app/assets/javascripts/admin/addon/templates/flags-index.hbs .': Invalid argument`

I'm using ubuntu
